app/models/manager_refresh/save_collection/helper.rb

```diff
@@ -14,85 +14,80 @@ def save_inventory_object_inventory(ems, inventory_collection)
-        inventory_collection.each do |inventory_object|
-          hash   = inventory_object.attributes(inventory_collection)
-          record = record_index.delete(inventory_object.manager_uuid)
-          if record.nil?
-            next unless inventory_collection.create_allowed?
-            record = inventory_collection.model_class.create!(hash.except(:id))
-            created_counter += 1
+        association.find_each do |record|
+          index = inventory_collection.object_index_with_keys(unique_index_keys, record)
+          if unique_db_indexes.include?(index) # Include on Set is O(1)
+            # We have a duplicate in the DB, destroy it. A find_each method does automatically .order(:id => :asc)
+            # so we always keep the oldest record in the case of duplicates.
+            _log.warn("A duplicate record was detected and destroyed, inventory_collection: '#{inventory_collection}', "\
+                      "record: '#{record}', duplicate_index: '#{index}'")
+            record.destroy
+          else
+            unique_db_indexes << index
+          end
+
+          inventory_object = inventory_objects_index.delete(index)
+          hash             = attributes_index.delete(index)
+
+          if inventory_object.nil?
+            next unless inventory_collection.delete_allowed?
+            deleted_counter += 1
+            record.public_send(inventory_collection.delete_method)
           else
             record.assign_attributes(hash.except(:id, :type))
             if inventory_collection.check_changed?
               record.save! if record.changed?
             else
               record.save!
             end
+
+            inventory_object.id = record.try(:id)
           end
-          inventory_object.id = record.try(:id)
         end
       end
-      _log.info("*************** PROCESSED #{inventory_collection}, created=#{created_counter}, "\
-                "updated=#{inventory_collection_size - created_counter} ***************")
-    end
 
-    def delete_inventory!(inventory_collection, record_index, association)
-      # Delete the items no longer found
-      unless record_index.blank?
-        deletes = record_index.values
-        _log.info("*************** DELETING #{inventory_collection} of size #{deletes.size} ***************")
-        type = association.proxy_association.reflection.name
-        _log.info("[#{type}] Deleting with method '#{inventory_collection.delete_method}' #{log_format_deletes(deletes)}")
+      if inventory_collection.create_allowed?
         ActiveRecord::Base.transaction do
-          deletes.map(&inventory_collection.delete_method)
+          inventory_objects_index.each do |index, inventory_object|
+            hash            = attributes_index.delete(index)
+            record          = inventory_collection.model_class.create!(hash.except(:id))
+            created_counter += 1
+
+            inventory_object.id = record.try(:id)
```

Comment:
the .try was there because of the old refresh (took it from there). We don't do create! and update! there, so it can happen the record is not created.